Lib/nntplib.py

```diff
@@ -325,7 +330,21 @@ def __init__(self, file, host,
         readermode.
         """
         self.host = host
-        self.file = file
+        self.port = port
+        self.sock = self._create_socket(timeout)
+        self.file = None
+        try:
+            self.file = self.sock.makefile("rwb")
+            self._base_init(readermode)
+            if user or usenetrc:
+                self.login(user, password, usenetrc)
+        except:
+            if self.file:
+                self.file.close()
+            self.sock.close()
+            raise
+
+    def _base_init(self, readermode):
```

Comment:
Does `_base_init` need to be public? I notice it's used in the test, but would it be expected to be used by non-test code?

Reply:
I agree that the changed structure is not that good.
Do you have any suggestions to be better in the abstract?

Reply:
Well, normally I'd say add an intermediate base class with a different constructor! :)

I don't think the code is so bad, but I'm questioning if code other than test code needs the same pattern, and if so, we should make this public.

Reply:
@ericvsmith 
My first intension was removing the intermediate base class :)
See: https://github.com/python/cpython/pull/17939#pullrequestreview-341290152

Unfortunately, the test code is using the base class.
So I choose this usage and _NNTPBase was not the public class.

Reply:
So the root of the problem is that the test code wants to pass in (or set) the `file` parameter when initializing the `NNTP` object. If that's something that we generally we don't want to support, then I think using the private `_base_init` is okay for the test code. It's a little odd that `NNTP.__init__` is never being called, and `self.port` is never being initialized, but I guess that's okay for the unsupported test case code.

So, after thinking about it: I'd make some comment in `_base_init` about why it exists (only to support testing where `file` is passed in), and not change any of the actual code here.